### ISSUE TITLE
Fix device error reporting

### ISSUE THREAD

**Issue by killenb** [ChimeraTK/ApplicationCore#53]
The device error reporting scheme is broken. As requested in #11 / #12 , reportError() blocks until the device error is resolved. This is conceptually wrong for the following reasons:
 * Writing must always succeed. Due to inversion of control a module cannot know whether it is connected to a device and would block (as mentioned as "paused" in #10) or if there is a fan-out in between and the module does not even know that the device blocked.
* If a module is blocked it cannot propagate the data invalidity flag, and not even shut down.
Currently, if the device stays in an error mode, it blocks the control system module if directly connected to the device, and prevents shutdown of the application. The only reason that this has not hit us so far is that the DeviceModule thread does not check if the device is back online (not in the interface yet), just sleeps 0.5 seconds and then always resets the error state.

Task:
 * [x] Extend the DeviceAccess Device and Backend interfaces with bool getDeviceError(), so a backend can signal a device error. By default, this reports falser, so backends do not have to implement this (some maybe even can't).
 * [x] Change the ExceptionDevice so it overloads getDeviceError and reports true if it would throw an exception.
 * [x] Change the DeviceModule so it keeps polling the DeviceError flag until it goes to true. Make sure there is an interuption point so the thread can terminate if the device does not become available again. The thread must not get stuck waiting for this flag, but must keep reporting errors that are reported (FIXME: does this make sense? The current implementation does not even report errors, so we might as well wait).
 * [x] Fix the implementation until all the broken and stuck tests work again. First in the master (part of this ticket), then in the issue46 branch (ticket #46 postponed until this one is done).
 * [ ] Update the documentation with the changed behaviour.
 * [x] Needs review by Martin K.

Consider the following for changes in the ExceptionHandlingDecorator:
* Writing must never block
* Read may block, but it might be better to return with invalid data
* readLatest/ readNonBlocking must not block but return with invalid data

**Comment by mhier** [comment]
A few comments: Device::open() can be called always (even if the device is already open) and shall throw an exception if the Device is in an exception state. Maybe backends need to be updated for this, but getDeviceError() is not needed then.
Then a Application Module blocked by an exception needs to report the invalidity flag only once, which should be done before blocking the module. I don't think this requires any structural change.
Also blocking writes are in principle ok in case of errors, since the module already is dysfunctional due to the device exception. I am pretty sure if this is not the intended behaviour in a particular module, the granularity of the module is wrong. By definition, modules shall have a single task only.

**Comment by killenb** [comment]
Thanks for the hint with the open() function. Is this behaviour mentioned in the specs/docu? If not we should mention it. (Although it seems a bit strange that isOpened() reports true, and then I call open() and it throws. Asking for an error state seems more like "least surprise" interface).

Currently a blocking write in the tests for #46 is a show stopper. The control system module is dysfunctional. I do not get updates of the error messages from the DeviceModule, and I cannot call the Application destructor because the thread does not join.

How could a module, that freezes on a write to some variable, update it's outputs with the invalid flag?
I think allowing writes to block completely contradicts our design with lockfree queues, which were supposed to make sure that threads are not blocked (because due to inversion of control you never know what you are connected to). The default behaviour of a module that does not explicitly check for data validity is to just keep running and attach the invalid flag  (see requirements in #38)

**Comment by mhier** [comment]
It's obviously insufficiently documented...

I think I understand the problem now. We should differentiate the behaviour of the application interface and the internal interface. There must be a way to write to the output which does not block to propagate the error, but this does not mean that the application module does not block in a write().

**Comment by mhier** [comment]
Btw: an error in a write() should not be propagated as a data invalid flag to other outputs.

Also: if a write never blocks, the unsuccessful write would be silently ignored. I think this is not an acceptable behaviour.

**Comment by killenb** [comment]
I still don't see why a module should block on an output.

Example: A module has two inputs and from this it computes two outputs. The inputs are coming from device A, one output goes to the CS, the other one to device B. Why should writing to B block the module? And if I now connect the second output also to the CS, the fan output is what blocks, and the module runs through. This is inconsistent.

I agree that write errors should not generate a data invalid flag in the module (in the example a above all inputs from Device A are valid, so the data is valid.) But output errors can only be ignored. Writing is "fire and forget". That's how we designed it. Otherwise we would have to propagate write errors back through the fan-outs, and then what could the application module do differently if all its inputs are still OK?

I think the right behaviour (and I believe it's implemented like that) is that a failing write does nothing to the sending module, but the device reports an error to the control system.

**Comment by mhier** [comment]
Well, if write() does not block, one would have to automatically retry the failed write. This will be very complicated to implement, especially if we want to keep the correct order of multiple failed writes across different variables. This is much more important than anything else. If we can guanrantee the success at least of the latest write for each variable, we can make write non-blocking, I ageee. 

But acutually I might prefer resolving the inconsistency in the other direction: we could make the write blocking even if a FanOut is involved. This would also make the behaviour of the application more consistent: I don't want to see updates of the dysfunctional module in the control system, because this suggests a state which is simply untrue.

**Comment by mhier** [comment]
Besides, I don't think the asymmetry is real: FanOuts for outputs are not threaded, so if any output of that FanOut blocks so does the entire write of the FanOut.

**Comment by killenb** [comment]
Not seeing repeated updates of invalid variables indeed is something we should have. But I don't think blocking writes is the correct implementation.

How about solving this inside the ProcessArray. If the dataInvalid flag is set, the value is not updated, and the old value is send once (do we have the old value?), then writes just don't do anything. Like this threads that receive data are not updated any more, incl. the control system. The worst that can happen is that a valid trigger starts a module which only does read_latest() and updates its invalid outputs, which then are not further distributed.

**Comment by killenb** [comment]
The blocking solution has one issue I see with the implementation: one has to be able to stop the application when the device is not working. For this I have to send interruption signals to all threads in all modules (and threaded fansouts etc), and then call the notify of all condition variables. And this is before I call the destructors of any module because they block each other. In a test I have the situation that the destructor of the device module, which would notify the condition variables, is not even called because the module destructor before it never returns. I have to externally kill the process.

**Comment by killenb** [comment]
In addition I think your argument about sending multiple variables and re-sending on failure does not hold.

Lets assume I send two variables in order, and the first one works, the seconds one fails. The sending blocks until the device becomes available again. The device has been rebooted and lost all values. The first thing that will happen when it becomes available is the second variable being written without the first one. Exactly the opposite of what you wanted.

I think this cannot be solved on the sender side. In application core you can use data consistency groups.  But this ticket concerns devices. If a device really requires an order, it can only assure this by a handshake sequence (write "start transaction" register, send the data words, write "end tansaction" register). Only if the sequence is complete the data is applied. Application core conceptually cannot solve it.

**Comment by mhier** [comment]
I think you are mixing up a few things. This has nothing to do with invalidity flags (which are not set after failed writes). Also, after a failure a device doesn't necessarily need to be power cycled, it could be just a tenporary network failure. 

The idea is that a module which cannot operate due to an exception state should be paused until the problem is fixed. I think blocking a write operation until it can succeed fits well into this scheme.

Still most importantly, we need the guarantee that writes are successful sooner or later. This is not something we should drop just because some internal Implementation might then be easier (and this is what triggered this discussion right now). Also the order across variables is important. Think about the consequences if you don't have that when you have an udp-based protocol over an unstable Ethernet connection: you simply wouldn't know which write has actually an effect. This is a very bad idea. 

Ps: we already guarantee the order of writes now, without any transaction or so. There is no need to break that. Maybe we could later introduce transactions, to make sure a block is consistently transmitted in case of an error, but this should be required in some special cases.

**Comment by mhier** [comment]
About the shutdown issue: please have a look at the implementation for read(). There is an easy solution.

**Comment by killenb** [comment]
It seems we have to get back to the drawing board and exactly define the wanted behaviour.
Temporary connection failures are a different issue. The backend has to take care about re-sending. And obviously data must be send in sequence, this never was in question. 

But I still think application modules conceptually cannot solve this. They are local code which does not care what it's connected to. That's the abstraction we made and that;'s what's making application core strong.

In my opinion blocking writes is a no-go. Take my example where all inputs are OK and only one output is cabled to one broken device, the other outputs are connected to working modules. Would you want to stop the whole module just because a status cannot be displayed on a device?
This obviously strongly depends on the use case.

What is also bad about blocking: You cannot react on errors. Even if the local code knows what to do, it cannot because it is just blocked.

**Comment by mhier** [comment]
I agree. We should split off this discussion from this ticket. This decision should not be driven by internal needs.

The current situation is that we already have blocking writes. The write function will block until the operation is complete or failed, which might take even seconds or minutes in case of network timouts. This behaviour shall not be changed within this ticket.

I think there are some misunderstandings about the current situation. In an application module we by definition never see exceptions, so we cannot react to errors at all. The backend doesn't handle errors - it throws exceptions to notify about errors. Hence this error must be handled by ApplicationCore one way or the other. This is a strong requirement.

**Comment by killenb** [comment]
Ah, and it's not only about shutdown. The control system module is already frozen during the tests to an extend that I cannot even perform the tests. It freezes before I can even get the device status in the test sequence. And it freezes so hard that I cannot even readNonBlocking(), because it t blocks. This shows that something is conceptually wrong.

**Comment by killenb** [comment]
OK, where to we put the discussion about the error propagation in AppicationCore?

**Comment by mhier** [comment]
Coming back to this ticket: there are solutions to the descibed problems which do not involve changing the interface of the write() function used by user code. Please have a look at the implementation for read(), most of these issues are solved there. When propagating data fault flags (due to an exception in a read()), one needs to call a different write function than the user code calls. 

Maybe one should introduce something like a writeNonBlocking() (maybe wiht a better Name), which would also give the user the choice whether a write action should be blocking or might be unreliably.

**Comment by mhier** [comment]
About the tests: probably the tests are wrongly designed.

**Comment by killenb** [comment]
I was thinking the same about the writeNonBlocking. But which one to chose in "everyday code"?
With blocking read() vs. readLatest() this is clear to me. But with the writes?

Besides, I noticed that the problem might be that readNonBlocking is blocking. It should not. it should return (and have the data invalid flag set).

**Comment by killenb** [comment]
The test design is rather simple:
- Start the app
- Make the ExceptionDevice fail on write.
- Write a cs to device variable (at this point everything freezes)
- Read the device status (which never returns because readLatest() is blocking)

The reason why the other exception tests are not blocking is because the DeviceModule always unconditionally resets the DeviceStatus to OK after 500 ms (which this ticket is supposed to fix).

I think I have to split this ticket into several. Not sure how, though. I will have to implement the check if the device is OK again in the module thread and see what breaks before I can continue.

**Comment by killenb** [comment]
Ok, this is strange. The tests run through now, except for the shutdown problem (which is solvable).

**Comment by mhier** [comment]
> I was thinking the same about the writeNonBlocking. But which one to chose in "everyday code"?
> With blocking read() vs. readLatest() this is clear to me. But with the writes?

Usually it doesn't matter. If the module runs periodically anyway, reliability is not really an issue. Also most modules write their stuff only to a single device (and maybe the control system), in which case the blocking implementation has only advantages (no fake operation when the target device is not available). Only in special cases the non-blocking write might be required. The only case I see is when there is one module which writes to two devices, which are highly asymmetric. One device is critical and the other e.g. only needs to be informed about some status. In this case writing to the non-critical device might be non-blocking (and hence unreliably), to ensure that the critical device gets its data even if the non-critical device is down.

If the unreliable non-blocking write would be the default, one would have easily inconsistencies between the control system and the actual machine state during a communication down time. Also: the current write implementation (even already in DeviceAccess) is defacto blocking. Changing this might break a lot of applications.

> Besides, I noticed that the problem might be that readNonBlocking is blocking. It should not. it should return (and have the data invalid flag set).

Yes this is true, this should be changed.

Regarding the test: if the test is broken it is more about how these steps are actually implemented, i.e. who reads what etc. Reading the device status itself shouldn't block, since this is a control system PV. Maybe you are trying to do too many things in the ApplicationModule thread? Of course the tests can well be wrong for this ticket, I would expect that they need to be changed.

**Comment by killenb** [comment]
There is no application module thread. I only have a device connected to the CS. Simplest possible application.
 And this freezes the CS thread so that readNonBlocking in fact was blocking. Or at least yesterday it was, and today it is not??? I have to dig what is going on.

**Comment by mhier** [comment]
Ah that need to be fixed differently!

**Comment by mhier** [comment]
Wait a second: why does the control system thread block? There should be a ThreadedFanOut for a direct CS-to-Device variable. This FanOut should be blocked. The control system thread always only sees the ProcessArray.

**Comment by killenb** [comment]
I don't know what blocked yesterday. I cannot reproduce it. Currently all tests run through. In spite of my suspicion that introducing the correct behaviour of the DeviceModule thread (does not resolve the error status while the Device reports errors), none of the existing tests broke.

**Comment by killenb** [comment]
The only thing that does not work yet is shutting read() down the application with a failing device and blocked accessors. There is no test for it, but if comment out the one line that "repairs" the ExceptionDummy in the tests, the test does not finish (blocked in the destructor).

**Comment by killenb** [comment]
BTW. Calling open() again did not work out. Device::open() does not even call the backend open() if the backend was opened, I think to mask that some backends throw() if open is called again (or were throwing when the code was implemented like that). I was afraid to open a can of worms if I touched that, and did not want to risk it now where I have to fight the strange hanging accessors.

In addition the calling code with` while(true){ try{ sleep(); open(); break;} catch(){}}` looked ugly,  and very much like misusing the open function, incl. building business logic on exceptions.

I introduced a function isFunctional(). It returns true if the device/backend is working fine, and false if it's closed, not connected or has errors. Now the code reads: `while( !device.isFunctional() ){sleep();}` I like that much better.

**Comment by mhier** [comment]
I find the name a bit misleading. The backend needs to perform a potentially expensive action to return the result. IsFunctional() sounds like returning a known status flag. Maybe tryRecover() or recoverFromException() or so would give the better picture?

**Comment by killenb** [comment]
I agree the name is not ideal. The original idea was something like `hasError()`, but what to return if the device has not been opened? What I want to know is if I can successfully read or write or will get an exception if I try. Or short: I want to know if that thing is working as intended (yes or no), so we came up with the name `isFunctional()`. Until now we assume that the device is functional when it has been opened, and the exceptions are indications that this assumption is wrong.

However, like with a theory, which cannot be proven correct but can only be proven wrong, the answer to `isTheoryCorrect()` is not _no_ and _yes_, but _no_ and _maybe_. The correct question is `isTheoryFalsified()`, which can be anwered with _yes_ and _no_. We need the equivalent for the device.

`isKnownToBeDysfunctional()` would be correct (`true` if there are known errors, if closed or disconnected, false otherwise), but the name is anything but catchy and intuitive. Very academic.

I don't know what tryRecover() and recoverFromException() is supposed to mean. That the device is currently trying to recover? But what if it has not even been opened? If it is supposed to mean the the code using the device should recover it is a "birthday opener". Don't name it what you use it for but what it does/is.

**Comment by mhier** [comment]
My point was: it is an action not a simple getter. Calling the function will (hopefully) change the internal state of the Device object.

**Comment by killenb** [comment]
Oh, this is a misunderstanding. It is a simple getter. It only checks if the device has seen errors and has been opened.

Now I start to understand why you wanted to use the open() function. It should trigger a recovery mechanism in the device. I always assumed that the device would do that automatically in the background once it has detected an error, and then would reset the error flag. And the polling device module would eventually see that the device is working again.

**Comment by killenb** [comment]
So the next design question is: Does a device backend automatically recover and report a functional status, or does recovery have to be triggered by the application (ApplicationCore in our case)?

**Comment by mhier** [comment]
This would require an extra thread for each device. It was never forseen so far. There is a ticket in device access about the  recently changed requirement, and this foresees an automatic recovery only at the next action triggered by the application. I think this is sufficient, otherwise backends get even more complex to write.

**Comment by killenb** [comment]
This would mean I just have to keep trying to read/write, and the recovery would be "automatic". I have to check how well this plays with readNonBlocking, which does not retry. And I have to re-think the mechanism in ApplicationCore again, which anyway still has a problem. But maybe this ticket is obsolete then.

Although, having to call a dedicated recovery function (maybe open()) might make debugging and the program logic simpler. Not sure yet.

I will read through device access requirements. Than for the link.

**Comment by mhier** [comment]
For an application there is no difference at all by definition. Ä dedicated function (either open() or a new function) for recovery is preferrable since it will allow ApplicationCore to attempt a recovery before clearing the error flag (visible in the control system) - this is currently not solved properly.

We urgently need to write up specifications somehow in a central place for each library (Doxygen page)...

**Comment by killenb** [comment]
> About the shutdown issue: please have a look at the implementation for read(). There is an easy solution.

Could you be a bit more specific about this? Which read() exactly are you referring to? There is a shared state with a queue in UnidirectionalProcessArray::doReadTransfer(). Is this what you mean? What is the easy solution?

I am looking for something to release a condition variable (i.e. wake up the thread and acquire the lock so the interruption point can be effective) which lives in a decorator of an accessor that someone gave to me. And I don't know what else to use than a condition variable. I need the "notify_all" with an unknown number of "subscibers" (threads waiting for the condition var)

**Comment by killenb** [comment]
Hah, that might solve it. The decorator must override transferFuture::interrupt(). It knows about the condition variable.

**Comment by mhier** [comment]
Sorry I thought this would have been implemented like this for the reading case already....

**Comment by killenb** [comment]
I have it solved now. There were basically three places where it blocked:
- The decorator had to notify the condition variable
- The Fan-Outs did not call interrupt on the slaves, only on the impl. (and the threaded fan had to notify the condition variable as well)
- The ApplicationModule suppressed calling interrupt on feeding accessors, but these now can also block.
It is not merged into the trunk yet because I have to clean up the debug printout. One cannot work in testable mode and use its debug printing.

**Comment by killenb** [comment]
I will make a ticket which asks you to review the test. I tried to cover as much as possible (directly conneting CS to device, directly module to device and the connection with triggered and untriggered fans. But I don't know what to do to cover FeedingFanOuts and Consumin and Threaded etc. For instance testBiDirectional showed an issue which my tests did not cover yet at that time. Don't know if it does now.

**Comment by killenb** [comment]
The ticket could be solved without non-blocking write. No behaviour change, so nothing to document.
